File: src/Array/SpiralMatrix/SpiralMatrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {        
        //Define borders
        int m = matrix.length, n = matrix[0].length;
        int left = 0, right = n - 1, top = 0, down = m - 1;
        
        //Define result arr
        Integer[] res = new Integer[m * n];
        int index = 0;
        
        while(index < res.length){
            //Traverse right
            for(int i = left; i <= right && index < res.length; i++){
                res[index++] = matrix[top][i];
            }
            top++;
            //Traverse down
            for(int i = top; i <= down && index < res.length; i++){
                res[index++] = matrix[i][right];
            }
            right--;
            //Traverse left
            for(int i = right; i >= left && index < res.length; i--){
                res[index++] = matrix[down][i];
            }
            down--;
            //Traverse top
            for(int i = down; i >= top && index < res.length; i--){
                res[index++] = matrix[i][left];
            }
            left++;
        }
        return Arrays.asList(res);
    }
}

File: src/Array/SpiralMatrix/SpiralMatrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        row_begin = 0
        col_begin = 0
        row_end = len(matrix)-1 
        col_end = len(matrix[0])-1
        total = len(matrix) * len(matrix[0])
        while len(res) < total:
            for i in range(col_begin,col_end+1):
                res.append(matrix[row_begin][i])
            row_begin += 1
            for i in range(row_begin,row_end+1):
                res.append(matrix[i][col_end])
            col_end -= 1
            if (row_begin <= row_end):
                for i in range(col_end,col_begin-1,-1):
                    res.append(matrix[row_end][i])
                row_end -= 1
            if (col_begin <= col_end):
                for i in range(row_end,row_begin-1,-1):
                    res.append(matrix[i][col_begin])
                col_begin += 1
        return res
